contracts/cw-names-registry/src/tests.rs

```diff
@@ -0,0 +1,477 @@
+    )
+    .unwrap();
+
+    // Try to reserve registered name, will fail
+    reserve(
+        &mut app,
+        names.clone(),
+        already_registered_name.to_string(),
+        Addr::unchecked(ADMIN_ADDR),
+    )
+    .unwrap_err();
+
+    // Try to reserve as non admin, will fail
+    reserve(
+        &mut app,
+        names.clone(),
+        name.to_string(),
+        Addr::unchecked(NON_ADMIN_ADDR),
+    )
+    .unwrap_err();
+
+    let res = query_name(&mut app, names.clone(), name.to_string());
+    assert!(!res.reserved);
+
+    // Reserve a name
+    reserve(
+        &mut app,
+        names.clone(),
+        name.to_string(),
+        Addr::unchecked(ADMIN_ADDR),
+    )
+    .unwrap();
+
+    let res = query_name(&mut app, names.clone(), name.to_string());
+    assert!(res.reserved);
+
+    // Try to transfer as non admin
+    transfer_reservation(
+        &mut app,
+        names.clone(),
+        name.to_string(),
+        DAO_ADDR.to_string(),
+        Addr::unchecked(NON_ADMIN_ADDR),
+    )
+    .unwrap_err();
+
+    // Try to transfer unreserved name
+    transfer_reservation(
+        &mut app,
+        names.clone(),
+        "NotReserved".to_string(),
+        DAO_ADDR.to_string(),
+        Addr::unchecked(ADMIN_ADDR),
+    )
+    .unwrap_err();
+
+    // Try to transfer to a DAO that already has a name
+    transfer_reservation(
+        &mut app,
+        names.clone(),
+        name.to_string(),
+        NON_ADMIN_ADDR.to_string(),
+        Addr::unchecked(ADMIN_ADDR),
+    )
+    .unwrap_err();
+
+    // Successfully transfer
+    transfer_reservation(
+        &mut app,
+        names.clone(),
+        name.to_string(),
+        DAO_ADDR.to_string(),
+        Addr::unchecked(ADMIN_ADDR),
+    )
+    .unwrap();
+
+    let res = query_name(&mut app, names.clone(), name.to_string());
+    assert!(!res.reserved);
+    assert_eq!(res.dao, Some(Addr::unchecked(DAO_ADDR)));
+
+    // Try to reserve the newly registered name from transfer, will fail
+    reserve(
+        &mut app,
+        names,
+        name.to_string(),
+        Addr::unchecked(ADMIN_ADDR),
+    )
+    .unwrap_err();
+}
+
+#[test]
+fn test_update_config() {
+    let mut app = App::default();
+    let (names, token) = setup_test_case(&mut app, Uint128::new(50));
+    let other_token = create_token(&mut app); // To be used when updating payment token
+
+    let config = query_config(&mut app, names.clone());
+    assert_eq!(
+        config,
+        Config {
+            admin: Addr::unchecked(ADMIN_ADDR),
+            payment_token_address: token,
+            payment_amount: Uint128::new(50),
+        }
+    );
+
+    // Update config as non admin fails
+    update_config(
+        &mut app,
+        names.clone(),
+        Some(other_token.to_string()),
+        Some(NON_ADMIN_ADDR.to_string()),
+        Some(Uint128::new(25)),
+        Addr::unchecked(NON_ADMIN_ADDR),
+    )
+    .unwrap_err();
+
+    // Update config as admin
+    update_config(
```

Comment:
See comment above. I'm not sure why this makes me so nervous, but whenever there are lots of `Option<>`s and conditional updating, I get a bit sweaty. It would be good to sense check that these update correctly when we're only setting one field at a time.